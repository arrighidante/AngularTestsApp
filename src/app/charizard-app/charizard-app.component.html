<h1>Pokemon</h1>
<hr />

@if(!charizard){
<div>
  <h2>Loading...</h2>
</div>
} @else {
<h3>{{ charizard.name | titlecase }}</h3>
<img [src]="charizard.sprites.front_default" [alt]="charizard.name" />
}

<form [formGroup]="clientForm">
  <input type="text" formControlName="name" />
  <input type="text" formControlName="email" />
  <input type="text" formControlName="password" />
  <button type="submit" (click)="submitForm()">Test</button>
</form>
