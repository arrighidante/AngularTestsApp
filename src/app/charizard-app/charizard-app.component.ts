import { TitleCasePipe } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { Component, OnInit } from '@angular/core';
import {
  FormControl,
  FormGroup,
  ReactiveFormsModule,
  Validators,
} from '@angular/forms';
import { PokemonService } from '../services/pokemon.service';
import { Pokemon } from './interfaces';

@Component({
  selector: 'app-charizard-app',
  standalone: true,
  imports: [TitleCasePipe, HttpClientModule, ReactiveFormsModule],
  templateUrl: './charizard-app.component.html',
  styleUrl: './charizard-app.component.scss',
})
export class CharizardAppComponent implements OnInit {
  constructor(private pokemonService: PokemonService) {}
  public charizard?: Pokemon;

  clientForm = new FormGroup({
    name: new FormControl('', [Validators.minLength(4)]),
    email: new FormControl('', [Validators.email, Validators.required]),
    password: new FormControl('', [
      Validators.required,
      Validators.minLength(4),
    ]),
  });

  ngOnInit(): void {
    this.pokemonService.getPokemon(6).subscribe((pokemon) => {
      this.charizard = pokemon;
    });
  }

  submitForm() {
    console.log(this.clientForm);
  }
}
