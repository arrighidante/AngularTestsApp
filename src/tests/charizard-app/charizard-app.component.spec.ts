import {
  HttpClientTestingModule,
  HttpTestingController,
} from '@angular/common/http/testing';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { CharizardAppComponent } from '../../app/charizard-app/charizard-app.component';
import { PokemonService } from '../../app/services/pokemon.service';

describe('CharizardAppComponent', () => {
  let component: CharizardAppComponent;
  let fixture: ComponentFixture<CharizardAppComponent>;
  let compiled: HTMLElement;
  let service: PokemonService;
  let httpMock: HttpTestingController;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [],
      imports: [HttpClientTestingModule, CharizardAppComponent],
      providers: [PokemonService],
    }).compileComponents();

    fixture = TestBed.createComponent(CharizardAppComponent);
    component = fixture.componentInstance;
    service = TestBed.inject(PokemonService);
    httpMock = TestBed.inject(HttpTestingController);

    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  afterEach(() => {
    httpMock.verify();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(compiled.innerHTML).toMatchSnapshot();
  });

  it('should show loading when the app starts', () => {
    const h2 = compiled.querySelector('h2');
    expect(h2?.textContent).toContain('Loading...');
  });

  it('should charizard data immediately after the component is created', () => {
    const dummyPokemon = {
      name: 'charizardo',
      sprites: {
        front_default: 'https://charizard.com/sprite.png',
      },
    };

    const url = 'https://pokeapi.co/api/v2/pokemon/6';

    // THIS SHOULDN'T BE HERE Since I'm testing only the component, not the service.
    service.getPokemon(6).subscribe((pokemon) => {
      expect(pokemon).toEqual(dummyPokemon);
    });

    const request = httpMock.expectOne(url);
    expect(request.request.method).toBe('GET');

    request.flush(dummyPokemon);

    fixture.detectChanges();
    console.log(compiled.innerHTML);
    // expect(component.charizard).toEqual(mockPokemon);
    const h3 = compiled.querySelector('h3');
    const img = compiled.querySelector('img');

    // THIS IS NOT WORKING SINCE THE  MOCKED SERVICE IS NOT BEING CALLED
    // expect(h3?.textContent).toContain(dummyPokemon.name);
    // expect(img?.src).toBe(dummyPokemon.sprites.front_default);
    // expect(img?.alt).toBe(dummyPokemon.name);
  });
});
