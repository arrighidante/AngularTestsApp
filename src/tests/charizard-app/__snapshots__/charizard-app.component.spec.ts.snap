// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharizardAppComponent should match the snapshot 1`] = `"<h1>Pokemon</h1><hr><div><h2>Loading...</h2></div><!--container--><!--container-->"`;
